common/lib/xmodule/xmodule/util/adaptive_learning.py

```diff
@@ -0,0 +1,270 @@
+        try:
+            link = next(
+                l for l in links if l.get('knowledge_node_uid') == block_id and l.get('student_uid') == user_id
+            )
+        except StopIteration:
+            link = None
+        return link
+
+    def get_knowledge_node_students(self):
+        """
+        Return list of all 'knowledge node student' objects for this course.
+        """
+        url = self.knowledge_node_students_url
+        response = requests.get(url, headers=self.request_headers)
+        links = json.loads(response.content)
+        return links
+
+    def create_knowledge_node_student(self, block_id, user_id):
+        """
+        Create 'knowledge node student' object that links student identified by `user_id`
+        to unit identified by `block_id`, and return it.
+        """
+        url = self.knowledge_node_students_url
+        payload = {'knowledge_node_uid': block_id, 'student_uid': user_id}
+        response = requests.post(url, headers=self.request_headers, data=payload)
+        knowledge_node_student = json.loads(response.content)
+        return knowledge_node_student
+
+    def create_knowledge_node_students(self, block_ids, user_id):
+        """
+        Create 'knowledge node student' objects that link student identified by `user_id`
+        to review questions identified by block IDs listed in `block_ids`, and return them.
+        """
+        knowledge_node_students = []
+        for block_id in block_ids:
+            knowledge_node_student = self.get_or_create_knowledge_node_student(block_id, user_id)
+            knowledge_node_students.append(knowledge_node_student)
+        log.error("KNOWLEDGE_NODE_STUDENTS: {}".format(knowledge_node_students))
+        return knowledge_node_students
+
+    def create_read_event(self, block_id, user_id):
+        """
+        Create read event for unit identified by `block_id` and student identified by `user_id`.
+        """
+        return self.create_event(block_id, user_id, 'EventRead')
+
+    def create_event(self, block_id, user_id, event_type):
+        """
+        Create event of type `event_type` for unit identified by `block_id` and student identified by `user_id`.
+        """
+        url = self.events_url
+        knowledge_node_student_id = self.get_knowledge_node_student_id(block_id, user_id)
+        payload = {
+            'knowledge_node_student_id': knowledge_node_student_id,
+            'event_type': event_type,
+        }
+        # Send request
+        response = requests.post(url, headers=self.request_headers, data=payload)
+        event = json.loads(response.content)
+        return event
+
+    def get_pending_reviews(self, user_id):
+        """
+        Return pending reviews for user identified by `user_id`.
+        """
+        url = self.pending_reviews_url
+        payload = {'student_uid': user_id}
+        response = requests.get(url, headers=self.request_headers, data=payload)
```

Comment:
@pomegranited Thanks for the detailed report :) I ran out of time for fixing these tests yesterday. They should pass as of a8e8dbc7a8121bdc0ff181d5671d0d29acf7450a.